// sprotty/lib/features/decoration/decoration-placer.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var smodel_1 = require("../../base/model/smodel");
var model_1 = require("./model");
var vnode_utils_1 = require("../../base/views/vnode-utils");
var geometry_1 = require("../../utils/geometry");
var model_2 = require("../bounds/model");
var model_3 = require("../routing/model");
var routing_1 = require("../routing/routing");
var DecorationPlacer = /** @class */ (function () {
    function DecorationPlacer() {
    }
    DecorationPlacer.prototype.decorate = function (vnode, element) {
        if (model_1.isDecoration(element)) {
            var position = this.getPosition(element);
            var translate = 'translate(' + position.x + ', ' + position.y + ')';
            vnode_utils_1.setAttr(vnode, 'transform', translate);
        }
        return vnode;
    };
    DecorationPlacer.prototype.getPosition = function (element) {
        if (element instanceof smodel_1.SChildElement && element.parent instanceof model_3.SRoutableElement) {
            var router = this.edgeRouterRegistry.get(element.parent.routerKind);
            var route = router.route(element.parent);
            if (route.length > 1) {
                var index = Math.floor(0.5 * (route.length - 1));
                var offset = model_2.isSizeable(element)
                    ? {
                        x: -0.5 * element.bounds.width,
                        y: -0.5 * element.bounds.width
                    }
                    : geometry_1.ORIGIN_POINT;
                return {
                    x: 0.5 * (route[index].x + route[index + 1].x) + offset.x,
                    y: 0.5 * (route[index].y + route[index + 1].y) + offset.y
                };
            }
        }
        if (model_2.isSizeable(element))
            return {
                x: -0.666 * element.bounds.width,
                y: -0.666 * element.bounds.height
            };
        return geometry_1.ORIGIN_POINT;
    };
    DecorationPlacer.prototype.postUpdate = function () {
    };
    __decorate([
        inversify_1.inject(routing_1.EdgeRouterRegistry),
        __metadata("design:type", routing_1.EdgeRouterRegistry)
    ], DecorationPlacer.prototype, "edgeRouterRegistry", void 0);
    DecorationPlacer = __decorate([
        inversify_1.injectable()
    ], DecorationPlacer);
    return DecorationPlacer;
}());
exports.DecorationPlacer = DecorationPlacer;
