// sprotty/lib/features/edit/create.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var command_1 = require("../../base/commands/command");
var smodel_1 = require("../../base/model/smodel");
var inversify_1 = require("inversify");
var types_1 = require("../../base/types");
var CreateElementAction = /** @class */ (function () {
    function CreateElementAction(containerId, elementSchema) {
        this.containerId = containerId;
        this.elementSchema = elementSchema;
        this.kind = CreateElementAction.KIND;
    }
    CreateElementAction.KIND = "createElement";
    return CreateElementAction;
}());
exports.CreateElementAction = CreateElementAction;
var CreateElementCommand = /** @class */ (function (_super) {
    __extends(CreateElementCommand, _super);
    function CreateElementCommand(action) {
        var _this = _super.call(this) || this;
        _this.action = action;
        return _this;
    }
    CreateElementCommand.prototype.execute = function (context) {
        var container = context.root.index.getById(this.action.containerId);
        if (container instanceof smodel_1.SParentElement) {
            this.container = container;
            this.newElement = context.modelFactory.createElement(this.action.elementSchema);
            this.container.add(this.newElement);
        }
        return context.root;
    };
    CreateElementCommand.prototype.undo = function (context) {
        this.container.remove(this.newElement);
        return context.root;
    };
    CreateElementCommand.prototype.redo = function (context) {
        this.container.add(this.newElement);
        return context.root;
    };
    CreateElementCommand.KIND = CreateElementAction.KIND;
    CreateElementCommand = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(types_1.TYPES.Action)),
        __metadata("design:paramtypes", [CreateElementAction])
    ], CreateElementCommand);
    return CreateElementCommand;
}(command_1.Command));
exports.CreateElementCommand = CreateElementCommand;
