// sprotty/lib/features/edit/delete.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var command_1 = require("../../base/commands/command");
var smodel_1 = require("../../base/model/smodel");
var types_1 = require("../../base/types");
var inversify_1 = require("inversify");
exports.deletableFeature = Symbol('deletableFeature');
function isDeletable(element) {
    return element instanceof smodel_1.SChildElement && element.hasFeature(exports.deletableFeature);
}
exports.isDeletable = isDeletable;
var DeleteElementAction = /** @class */ (function () {
    function DeleteElementAction(elementIds) {
        this.elementIds = elementIds;
        this.kind = DeleteElementAction.KIND;
    }
    DeleteElementAction.KIND = 'delete';
    return DeleteElementAction;
}());
exports.DeleteElementAction = DeleteElementAction;
var ResolvedDelete = /** @class */ (function () {
    function ResolvedDelete() {
    }
    return ResolvedDelete;
}());
exports.ResolvedDelete = ResolvedDelete;
var DeleteElementCommand = /** @class */ (function (_super) {
    __extends(DeleteElementCommand, _super);
    function DeleteElementCommand(action) {
        var _this = _super.call(this) || this;
        _this.action = action;
        _this.resolvedDeletes = [];
        return _this;
    }
    DeleteElementCommand.prototype.execute = function (context) {
        var index = context.root.index;
        for (var _i = 0, _a = this.action.elementIds; _i < _a.length; _i++) {
            var id = _a[_i];
            var element = index.getById(id);
            if (element && isDeletable(element)) {
                this.resolvedDeletes.push({ child: element, parent: element.parent });
                element.parent.remove(element);
            }
        }
        return context.root;
    };
    DeleteElementCommand.prototype.undo = function (context) {
        for (var _i = 0, _a = this.resolvedDeletes; _i < _a.length; _i++) {
            var resolvedDelete = _a[_i];
            resolvedDelete.parent.add(resolvedDelete.child);
        }
        return context.root;
    };
    DeleteElementCommand.prototype.redo = function (context) {
        for (var _i = 0, _a = this.resolvedDeletes; _i < _a.length; _i++) {
            var resolvedDelete = _a[_i];
            resolvedDelete.parent.remove(resolvedDelete.child);
        }
        return context.root;
    };
    DeleteElementCommand.KIND = DeleteElementAction.KIND;
    DeleteElementCommand = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(types_1.TYPES.Action)),
        __metadata("design:paramtypes", [DeleteElementAction])
    ], DeleteElementCommand);
    return DeleteElementCommand;
}(command_1.Command));
exports.DeleteElementCommand = DeleteElementCommand;
