// sprotty/lib/features/edit/reconnect.js
"use strict";
/********************************************************************************
 * Copyright (c) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var command_1 = require("../../base/commands/command");
var types_1 = require("../../base/types");
var model_1 = require("../routing/model");
var routing_1 = require("../routing/routing");
var ReconnectAction = /** @class */ (function () {
    function ReconnectAction(routableId, newSourceId, newTargetId) {
        this.routableId = routableId;
        this.newSourceId = newSourceId;
        this.newTargetId = newTargetId;
        this.kind = ReconnectAction.KIND;
    }
    ReconnectAction.KIND = 'reconnect';
    return ReconnectAction;
}());
exports.ReconnectAction = ReconnectAction;
var ReconnectCommand = /** @class */ (function (_super) {
    __extends(ReconnectCommand, _super);
    function ReconnectCommand(action) {
        var _this = _super.call(this) || this;
        _this.action = action;
        return _this;
    }
    ReconnectCommand.prototype.execute = function (context) {
        this.doExecute(context);
        return context.root;
    };
    ReconnectCommand.prototype.doExecute = function (context) {
        var index = context.root.index;
        var edge = index.getById(this.action.routableId);
        if (edge instanceof model_1.SRoutableElement) {
            var router = this.edgeRouterRegistry.get(edge.routerKind);
            var before = router.takeSnapshot(edge);
            router.applyReconnect(edge, this.action.newSourceId, this.action.newTargetId);
            var after = router.takeSnapshot(edge);
            this.memento = {
                edge: edge,
                before: before,
                after: after
            };
        }
    };
    ReconnectCommand.prototype.undo = function (context) {
        if (this.memento) {
            var router = this.edgeRouterRegistry.get(this.memento.edge.routerKind);
            router.applySnapshot(this.memento.edge, this.memento.before);
        }
        return context.root;
    };
    ReconnectCommand.prototype.redo = function (context) {
        if (this.memento) {
            var router = this.edgeRouterRegistry.get(this.memento.edge.routerKind);
            router.applySnapshot(this.memento.edge, this.memento.after);
        }
        return context.root;
    };
    ReconnectCommand.KIND = ReconnectAction.KIND;
    __decorate([
        inversify_1.inject(routing_1.EdgeRouterRegistry),
        __metadata("design:type", routing_1.EdgeRouterRegistry)
    ], ReconnectCommand.prototype, "edgeRouterRegistry", void 0);
    ReconnectCommand = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(types_1.TYPES.Action)),
        __metadata("design:paramtypes", [ReconnectAction])
    ], ReconnectCommand);
    return ReconnectCommand;
}(command_1.Command));
exports.ReconnectCommand = ReconnectCommand;
