// sprotty/lib/features/zorder/zorder.js
"use strict";
/********************************************************************************
 * Copyright (c) 2019 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var types_1 = require("../../base/types");
var smodel_1 = require("../../base/model/smodel");
var command_1 = require("../../base/commands/command");
var model_1 = require("../routing/model");
/**
 * Action to render the selected elements in front of others by manipulating the z-order.
 */
var BringToFrontAction = /** @class */ (function () {
    function BringToFrontAction(elementIDs) {
        this.elementIDs = elementIDs;
        this.kind = BringToFrontAction.KIND;
    }
    BringToFrontAction.KIND = 'bringToFront';
    return BringToFrontAction;
}());
exports.BringToFrontAction = BringToFrontAction;
var BringToFrontCommand = /** @class */ (function (_super) {
    __extends(BringToFrontCommand, _super);
    function BringToFrontCommand(action) {
        var _this = _super.call(this) || this;
        _this.action = action;
        _this.selected = [];
        return _this;
    }
    BringToFrontCommand.prototype.execute = function (context) {
        var _this = this;
        var model = context.root;
        this.action.elementIDs.forEach(function (id) {
            var element = model.index.getById(id);
            if (element instanceof model_1.SRoutableElement) {
                if (element.source)
                    _this.addToSelection(element.source);
                if (element.target)
                    _this.addToSelection(element.target);
            }
            if (element instanceof smodel_1.SChildElement) {
                _this.addToSelection(element);
            }
            _this.includeConnectedEdges(element);
        });
        return this.redo(context);
    };
    BringToFrontCommand.prototype.includeConnectedEdges = function (element) {
        var _this = this;
        if (element instanceof model_1.SConnectableElement) {
            element.incomingEdges.forEach(function (edge) { return _this.addToSelection(edge); });
            element.outgoingEdges.forEach(function (edge) { return _this.addToSelection(edge); });
        }
        if (element instanceof smodel_1.SParentElement) {
            for (var _i = 0, _a = element.children; _i < _a.length; _i++) {
                var child = _a[_i];
                this.includeConnectedEdges(child);
            }
        }
    };
    BringToFrontCommand.prototype.addToSelection = function (element) {
        this.selected.push({
            element: element,
            index: element.parent.children.indexOf(element)
        });
    };
    BringToFrontCommand.prototype.undo = function (context) {
        for (var i = this.selected.length - 1; i >= 0; i--) {
            var selection = this.selected[i];
            var element = selection.element;
            element.parent.move(element, selection.index);
        }
        return context.root;
    };
    BringToFrontCommand.prototype.redo = function (context) {
        for (var i = 0; i < this.selected.length; i++) {
            this.bringToFront(this.selected[i]);
        }
        return context.root;
    };
    BringToFrontCommand.prototype.bringToFront = function (selection) {
        var element = selection.element;
        var childrenLength = element.parent.children.length;
        element.parent.move(element, childrenLength - 1);
    };
    BringToFrontCommand.KIND = BringToFrontAction.KIND;
    BringToFrontCommand = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(types_1.TYPES.Action)),
        __metadata("design:paramtypes", [BringToFrontAction])
    ], BringToFrontCommand);
    return BringToFrontCommand;
}(command_1.Command));
exports.BringToFrontCommand = BringToFrontCommand;
